File: backend/src/models/ScheduledTask.ts
import mongoose, { Document, Schema } from 'mongoose';

export type TaskType = 'github_trending' | 'daily_poem' | 'daily_english' | 'custom';

export interface IScheduledTask extends Document {
    userId: mongoose.Types.ObjectId;
    taskType: TaskType;
    taskName: string; // 任务名称
    prompt?: string; // 自定义任务的提示词（仅 custom 类型需要）
    enabled: boolean;
    pushTime: string; // "HH:mm" format, e.g., "09:00"
    timezone: string; // IANA timezone, e.g., "Asia/Shanghai"
    conversationId?: mongoose.Types.ObjectId;
    nextRunAt?: Date | null;
    createdAt: Date;
    updatedAt: Date;
}

const scheduledTaskSchema = new Schema<IScheduledTask>(
    {
        userId: {
            type: Schema.Types.ObjectId,
            ref: 'User',
            required: true,
        },
        taskType: {
            type: String,
            enum: ['github_trending', 'daily_poem', 'daily_english', 'custom'],
            required: true,
        },
        taskName: {
            type: String,
            required: true,
        },
        prompt: {
            type: String,
        },
        enabled: {
            type: Boolean,
            default: false,
        },
        pushTime: {
            type: String,
            default: '09:00',
            validate: {
                validator: (v: string) => /^([01]\d|2[0-3]):([0-5]\d)$/.test(v),
                message: 'pushTime must be in HH:mm format',
            },
        },
        timezone: {
            type: String,
            default: 'Asia/Shanghai',
        },
        conversationId: {
            type: Schema.Types.ObjectId,
            ref: 'Conversation',
        },
        nextRunAt: {
            type: Date,
            default: null,
        },
    },
    {
        timestamps: true,
    }
);

// Compound index: one task type per user (for preset tasks)
// Custom tasks can have multiple per user
scheduledTaskSchema.index({ userId: 1, taskType: 1 }, { 
    unique: true,
    partialFilterExpression: { taskType: { $ne: 'custom' } }
});
scheduledTaskSchema.index({ userId: 1 });
scheduledTaskSchema.index({ enabled: 1, nextRunAt: 1 });

export default mongoose.model<IScheduledTask>('ScheduledTask', scheduledTaskSchema);

File: mobile/lib/data/models/scheduled_task.dart
class PresetTask {
  final String? id;
  final String taskType;
  final String taskName;
  final bool enabled;
  final String pushTime;
  final String? conversationId;

  PresetTask({
    this.id,
    required this.taskType,
    required this.taskName,
    required this.enabled,
    required this.pushTime,
    this.conversationId,
  });

  factory PresetTask.fromJson(Map<String, dynamic> json) {
    return PresetTask(
      id: json['_id'] as String?,
      taskType: json['taskType'] as String? ?? '',
      taskName: json['taskName'] as String? ?? '',
      enabled: json['enabled'] as bool? ?? false,
      pushTime: json['pushTime'] as String? ?? '09:00',
      conversationId: json['conversationId'] as String?,
    );
  }

  Map<String, dynamic> toJson() => {
    if (id != null) '_id': id,
    'taskType': taskType,
    'taskName': taskName,
    'enabled': enabled,
    'pushTime': pushTime,
    if (conversationId != null) 'conversationId': conversationId,
  };
}

class CustomTask {
  final String id;
  final String taskType;
  final String taskName;
  final String? prompt;
  final bool enabled;
  final String pushTime;
  final String? conversationId;

  CustomTask({
    required this.id,
    required this.taskType,
    required this.taskName,
    this.prompt,
    required this.enabled,
    required this.pushTime,
    this.conversationId,
  });

  factory CustomTask.fromJson(Map<String, dynamic> json) {
    return CustomTask(
      id: json['_id'] as String? ?? '',
      taskType: json['taskType'] as String? ?? 'custom',
      taskName: json['taskName'] as String? ?? '',
      prompt: json['prompt'] as String?,
      enabled: json['enabled'] as bool? ?? false,
      pushTime: json['pushTime'] as String? ?? '09:00',
      conversationId: json['conversationId'] as String?,
    );
  }

  Map<String, dynamic> toJson() => {
    '_id': id,
    'taskType': taskType,
    'taskName': taskName,
    if (prompt != null) 'prompt': prompt,
    'enabled': enabled,
    'pushTime': pushTime,
    if (conversationId != null) 'conversationId': conversationId,
  };
}

class TasksResponse {
  final List<PresetTask> presetTasks;
  final List<CustomTask> customTasks;

  TasksResponse({
    required this.presetTasks,
    required this.customTasks,
  });

  factory TasksResponse.fromJson(Map<String, dynamic> json) {
    return TasksResponse(
      presetTasks: (json['presetTasks'] as List<dynamic>?)
              ?.map((e) => PresetTask.fromJson(e as Map<String, dynamic>))
              .toList() ??
          [],
      customTasks: (json['customTasks'] as List<dynamic>?)
              ?.map((e) => CustomTask.fromJson(e as Map<String, dynamic>))
              .toList() ??
          [],
    );
  }

  Map<String, dynamic> toJson() => {
    'presetTasks': presetTasks.map((e) => e.toJson()).toList(),
    'customTasks': customTasks.map((e) => e.toJson()).toList(),
  };
}
